### PR TITLE
Property service API design

### LINKED ISSUE #615: Property-Service: API Design
Based on the MVP requirements of https://github.com/helpwave/helpwave/issues/109#issuecomment-1891115038, we need to design the corresponding API. The result of this issue should be a Profobuf definition, minor adjustments are possible during the actual implementation.

We need some management (CRUD) APIs for `Propery`, `PropertyValue` and `PropertySet`. `PropertyView` is in still in Todo. An integral part beside CRUD is are the APIs around appending new PropertyValues and reading the history of these values.

fyi: You can ignore our organization structure.

See #611 for Aggregates

### PR DESCRIPTION
**Pull request by FoseFx**
## Which issues does this pull request close?
closes #615

Adds:
- Set management
- Property management
- Attach Property Values
- Get Attached Property Values

TODO:
- views

TBD:
- filter / search params
- views and
- default list
- remove soft required
- allow custom values for selects
- move proto to its own service

### CODE REVIEW

**Comment by PaulKalho** [comment]
The `property-svc` will becom a seperate service distinct from the `tasks-svc`, why did you put the proto inside the `tasks-svc directory?`

File: proto/services/tasks_svc/v1/properties_svc.proto [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+//
+
+message GetAttachedPropertyValuesRequest {
+  string subject_id = 1; // @gotags: validate:"uuid4"
+
+  optional string view = 2;
+  // TODO: TBD: other filters / search params, e.g. include_is_archived
+}
+
+message GetAttachedPropertyValuesResponse {
```
Comment (MaxSchaefer) [inline proto/services/tasks_svc/v1/properties_svc.proto:229]: Based on https://github.com/helpwave/services/pull/617, this response will be quite hard to achieve without projections. The init PR will introduce `Property`, `PropertySet` and `PropertyValue` as separate gRPC-services.

`GetAttachedPropertyValues` can stay for now. Could you introduce `GetPropertyValue` that returns the scoped data of just the `PropertyValue` service?

**Review by MaxSchaefer — CHANGES_REQUESTED** [review]

File: proto/services/tasks_svc/v1/properties_svc.proto [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+//
+// GetProperties
+//
+
+message GetPropertiesRequest {
+  optional string set_id = 1; // @gotags: validate:"uuid4"
+
+  // TODO: TBD: other filters / search params, e.g. include_is_archived
```
[the remark this answers is not in the document]
Comment (FoseFx) [reply proto/services/tasks_svc/v1/properties_svc.proto:128]: should we do these now? Do we have a UI for this yet?

File: proto/services/tasks_svc/v1/properties_svc.proto [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+  SubjectType subject_type = 2; // TODO: add validation here (not-unspecified check)
+  FieldType field_type = 3; // TODO: add validation here (not-unspecified check)
+
+  string name = 4; // TODO: add validation (at least non-empty checks)
+  optional string description = 5; // TODO: add validation, if needed
+
+  // ID of set this Property should belong to
+  optional string set_id = 6;
+
+  optional bool always_include_for_current_context = 7;
```
Comment (FoseFx) [inline proto/services/tasks_svc/v1/properties_svc.proto:89]: In order to be able to do this, we need to get the context (above, proto value 1). For example "Ward".

File: proto/services/tasks_svc/v1/properties_svc.proto [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+  FIELD_TYPE_DATE = 4;
+  FIELD_TYPE_DATE_TIME = 5;
+  FIELD_TYPE_SELECT = 6;
+}
+
+// TODO: View"Context" is too general of a name for this in my opinion
+enum ViewContext {
+  VIEW_CONTEXT_UNSPECIFIED = 0;
+  VIEW_CONTEXT_GLOBAL = 1;
+  // TODO TBD: VIEW_CONTEXT_ORGANIZATION = 2;
```
[the remark this answers is not in the document]
Comment (MaxSchaefer) [reply proto/services/tasks_svc/v1/properties_svc.proto:26]:
```suggestion
```
Would be team

File: proto/services/tasks_svc/v1/properties_svc.proto [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+      string name = 2;
+    }
+    optional Set set = 7;
+
+    oneof value {
+      bool is_undefined = 8; // if it exists, it's true
+      string text_value = 9; // FIELD_TYPE_TEXT
+      float number_value = 10; // FIELD_TYPE_NUMBER
+      bool bool_value = 11; // FIELD_TYPE_CHECKBOX
+      string date_value = 12; // FIELD_TYPE_DATE TODO: is there a proto date type?
```
[the remark this answers is not in the document]
Comment (MaxSchaefer) [reply proto/services/tasks_svc/v1/properties_svc.proto:297]: https://github.com/googleapis/googleapis/blob/master/google/type/date.proto

File: proto/services/tasks_svc/v1/properties_svc.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaxSchaefer) [reply proto/services/tasks_svc/v1/properties_svc.proto]: We have a UI. But leave it as a TODO.
https://www.figma.com/file/4oxQRRLaNQ1TdBVhI0CZ4B/helpwave-frontend?type=design&node-id=2842-2158&mode=design&t=eVovSYHzSPim7EWW-4

File: proto/services/tasks_svc/v1/properties_svc.proto (hunk 3 above)
Comment (FoseFx) [reply proto/services/tasks_svc/v1/properties_svc.proto:89]: I think we need this, else the frontend would need to query the state of this button manually for every property:

image

Comment (MaxSchaefer) [reply proto/services/tasks_svc/v1/properties_svc.proto:89]: Okay, looks good. As mentioned on Slack, we need to talk about the PatientViews separately.

**Comment by FoseFx** [comment]
This PR and discussion should be used as reference for the implementation of #643

File: proto/services/properties_svc/v1/values.proto [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+message GetAttachedPropertyValuesResponse {
+  message Value {
+    string property_id = 1;
+    FieldType field_type = 2;
+    string name = 3;
+    optional string description = 4;
+
+    bool is_archived = 5;
```
Comment (FoseFx) [inline proto/services/properties_svc/v1/values.proto:49]: bool is_soft_required; missing

**Pull request closed without merging**